Convert products list to SQL

esoergel opened the pull request:
http://manage.dimagi.com/default.asp?163241
Pagination was super wonky, as we don't have access to archival status on the
couch view.  This meant the filtering happened in memory, so you wouldn't get
a full page of results.  It also appeared broken entirely when viewing
archived products.
@sravfeyn
@twymer FYI

esoergel committed:
Convert products list to SQL

Pagination was super wonky, as we don't have access to archival status
on the couch view.  This meant the filtering happened in memory, so you
wouldn't get a full page of results.  It also appeared broken entirely
when viewing archived products.

esoergel commented on corehq/apps/products/views.py:
```diff
@@ -113,51 +118,63 @@ def page_context(self):
-                kwargs={'domain': self.domain, 'prod_id': p._id}
-            )
-            data.append(info)
-        return data
```
This method was starting to get big with the switch to explicitly specifying all the fields sent over the wire, so I pulled it apart.

czue commented on corehq/apps/products/views.py:
```diff
@@ -76,20 +76,25 @@ class ProductListView(BaseCommTrackManageView):
     @property
-    def show_inactive(self):
-        return json.loads(self.request.GET.get('show_inactive', 'false'))
+    def show_only_inactive(self):
```
:+1: to clearer naming

czue commented on corehq/apps/products/views.py:
```diff
@@ -113,51 +118,63 @@ def page_context(self):
+
+    def program_name(self, product):
+        if product.program_id:
+            return Program.get(product.program_id).name
```
if you have 100 products with the same program ID this is going to make 100 couch calls. seems like it might be worth further tearing apart and caching it.

czue replied:
I don't think this needs to block merge especially since it's not a change in behavior - but I do think it's worth fixing it soon. Let me know if you want me to merge first and you'll track elsewhere.

esoergel replied:
Good call.  Just added https://github.com/dimagi/commcare-hq/commit/5e69c29daf08edb38ed67926146d494566e07a31 to avoid that.

czue commented:
yay sql!

esoergel committed:
Avoid duplicate program lookups

The lookup will still fail hard if the program isn't found, which is
consistent with the current behavior.  It shouldn't be the case that a
product has a non-existant program_id.  I wouldn't be the least
bit surprised if that does happen, and I think I'd rather have it fail
hard and obviously when it does.

czue commented:
:+1:
